src/Driver/GeckoDriver/Downloader.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DBrekelmans\BrowserDriverInstaller\Driver\GeckoDriver;
+
+use DBrekelmans\BrowserDriverInstaller\Archive\Extractor;
+use DBrekelmans\BrowserDriverInstaller\Driver\Downloader as DownloaderInterface;
+use DBrekelmans\BrowserDriverInstaller\Driver\Driver;
+use DBrekelmans\BrowserDriverInstaller\Driver\DriverName;
+use DBrekelmans\BrowserDriverInstaller\Exception\NotImplemented;
+use DBrekelmans\BrowserDriverInstaller\OperatingSystem\OperatingSystem;
+use RuntimeException;
+use Safe\Exceptions\FilesystemException;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+use function basename;
+use function dirname;
+use function Safe\fclose;
+use function Safe\fopen;
+use function Safe\fwrite;
+use function Safe\sprintf;
+use function strpos;
+use function sys_get_temp_dir;
+
+use const DIRECTORY_SEPARATOR;
+
+final class Downloader implements DownloaderInterface
+{
+    public const DOWNLOAD_PATH_OS_PART_WINDOWS = 'win64';
+    public const DOWNLOAD_PATH_OS_PART_MACOS = 'macos';
+    public const DOWNLOAD_PATH_OS_PART_LINUX = 'linux64';
+    public const DOWNLOAD_PATH = 'https://github.com/mozilla/geckodriver/releases/download/v%s/geckodriver-v%s-%s.tar.gz';
+
+    /** @var Filesystem  */
+    private $filesystem;
+
+    /** @var HttpClientInterface  */
+    private $httpClient;
+    /** @var Extractor */
+    private $archiveExtractor;
+
+    public function __construct(Filesystem $filesystem, HttpClientInterface $httpClient, Extractor $archiveExtractor)
+    {
+        $this->filesystem = $filesystem;
+        $this->httpClient = $httpClient;
+        $this->archiveExtractor = $archiveExtractor;
+    }
+
+    public function download(Driver $driver, string $location): string
+    {
+        try {
+            $archive = $this->downloadArchive($driver);
+        } catch (NotImplemented | FilesystemException | IOException | TransportExceptionInterface $exception) {
+            throw new RuntimeException('Something went wrong downloading the geckodriver archive.', 0, $exception);
+        }
+
+        try {
+            $binary = $this->extractArchive($archive);
```

Comment:
I created a ZipExtactor and a MultiExtractor that can handle several archive extractors.
I realised that I needed it also because archives are in zip format for windows and in tar.gz format for linux and macos.

I put also this MultiExtractor for handling archive extraction in ChromeDriver Downloader.